File: as1/savages/savages.go
package main

import (
    "fmt"
    "time"
    "sync"
)

type empty struct{}
type semaphore chan empty

func (s semaphore) acquire() {
    e := empty{}
    s<-e
}

func (s semaphore) release() {
    <-s
}

var CAPACITY = 20 // servings in a pot
var COOKS = 2
var COOK_QUOTA = 5000
var SAVAGES = COOKS * 4
var SAVAGE_QUOTA = (COOKS * COOK_QUOTA * CAPACITY) / SAVAGES

// primitives
var servings int = 0
var mutex = make(semaphore, 1)
var emptyPot = make(semaphore, 0)
var fullPot = make(semaphore, 0)

var exited = 0

func main() {
    startTime := time.Now()

    wg := new(sync.WaitGroup)
    wg.Add(COOKS + SAVAGES)

    for i := 0; i < COOKS; i++ {
        go cook(wg)
    }
    for i := 0; i < SAVAGES; i++ {
        go savage(wg)
    }

    wg.Wait()

    fmt.Printf("--- %s ---\n", time.Since(startTime))
}

func cook(wg *sync.WaitGroup) {
    defer wg.Done()

    for cooked := 0; cooked < COOK_QUOTA; cooked++ {
        emptyPot.acquire()
        fillPot()
        fullPot.release()
    }
}

func fillPot() {
    fmt.Println("Pot is now full")
}

func savage(wg *sync.WaitGroup) {
    defer wg.Done()

    for eaten := 0; eaten < SAVAGE_QUOTA; eaten++ {
        mutex.acquire()

        if (servings == 0) {
            emptyPot.release()
            fullPot.acquire()
            servings = CAPACITY
        }

        servings --
        getServing()

        mutex.release()

        fmt.Println("Savage is eating")
    }
}

func getServing() {
    fmt.Println("Savage retreived serving")
}

File: as1/savages/savages.py
from threading import Semaphore
from threading import Thread
import time

CAPACITY = 20 # servings in a pot
COOKS = 2
COOK_QUOTA = 5_000
SAVAGES = COOKS * 4
SAVAGE_QUOTA = int((COOKS * COOK_QUOTA * CAPACITY) / SAVAGES)

# primitives
servings = 0
mutex = Semaphore(1)
empty_pot = Semaphore(0)
full_pot = Semaphore(0)

def main():

    threads = []

    for i in range(COOKS):
        cook_t = Thread(target=cook, args=())
        cook_t.start()
        threads.append(cook_t)

    for i in range(SAVAGES):
        savage_t = Thread(target=savage, args=())
        savage_t.start()
        threads.append(savage_t)

    for thread in threads:
        thread.join()

def cook():
    for i in range(COOK_QUOTA):
        empty_pot.acquire()
        fill_pot(CAPACITY)
        full_pot.release()

def fill_pot(amount):
    print("Pot is now full")

def savage():
    global servings

    for i in range(SAVAGE_QUOTA):
        mutex.acquire()
        
        if servings == 0:
            empty_pot.release()
            full_pot.acquire()
            servings = CAPACITY

        servings -= 1
        get_serving()

        mutex.release()

        print("Savage is eating")

def get_serving():
    print("Savage retreived serving")

if __name__ == "__main__":
    start_time = time.time()
    main()
    print("--- %s seconds ---" % (time.time() - start_time))
